cudamapper/src/main.cu

```diff
@@ -177,38 +182,54 @@ int main(int argc, char* argv[])
+                                                                                                                            claragenomics::io::FastaParser& parser,
+                                                                                                                            const claragenomics::cudamapper::read_id_t start_index,
+                                                                                                                            const claragenomics::cudamapper::read_id_t end_index,
+                                                                                                                            const std::uint64_t k,
+                                                                                                                            const std::uint64_t w,
+                                                                                                                            const int device_id,
+                                                                                                                            const bool allow_cache_index,
+                                                                                                                            const double filtering_parameter) {
         CGA_NVTX_RANGE(profiler, "get index");
         std::pair<uint64_t, uint64_t> key;
         key.first  = start_index;
         key.second = end_index;
 
         std::shared_ptr<claragenomics::cudamapper::Index> index;
 
-        if (index_cache[device_id].count(key))
+        // first check if it's available on device, if not then check the host cache
+        if (device_index_cache[device_id].count(key))
+        {
+            index = device_index_cache[device_id][key];
+        }
+        else if (host_index_cache.count(key))
         {
-            index = index_cache[device_id][key];
+            index = host_index_cache[key]->copy_index_to_device(allocator);
         }
         else
         {
-            index = std::move(claragenomics::cudamapper::Index::create_index(allocator, parser, start_index, end_index, k, w));
+            index = std::move(claragenomics::cudamapper::Index::create_index(allocator, parser, start_index, end_index, k, w, true, filtering_parameter));
 
             // If in all-to-all mode, put this query in the cache for later use.
             // Cache eviction is handled later on by the calling thread
             // using the evict_index function.
-            if (get_size<int32_t>(index_cache[device_id]) < max_index_cache_size && allow_cache_index)
+            if (get_size<int32_t>(device_index_cache[device_id]) < max_index_cache_size_on_device && allow_cache_index)
             {
-                index_cache[device_id][key] = index;
+                device_index_cache[device_id][key] = index;
+            }
+
+            // update host cache
+            // ToDo: make it thread-safe when there are multiple GPUs
```

Comment:
Can you do this before the merge? We are trying to keep the code as multi-GPU-ready as possile

Reply:
the simplest way to keep it thread-safe was to limit write-access to only device 0. Alternatively could be done with mutex, but I doubt they make any significant difference performance-wise.